modules/optflow/src/sparse_matching_gpc.cpp

```diff
@@ -94,20 +117,321 @@ struct PartitionPredicate2
+  "__global const uchar* imgCh2, int ic2step, int ic2off,"
+  "__global uchar* out, int outstep, int outoff,"
+  "const int gh, const int gw, const int PR"
+  ") {"
+  "const int i = get_global_id(0);"
+  "const int j = get_global_id(1);"
+  "if (i >= gh || j >= gw) return;"
+  "__global double* desc = (__global double*)(out + (outstep * (i * gw + j) + outoff));"
+  "const int patchRadius = PR * 2;"
+  "float patch[" STR_EXPAND(PATCH_RADIUS_DOUBLED) "][" STR_EXPAND(PATCH_RADIUS_DOUBLED) "];"
+  "for (int i0 = 0; i0 < patchRadius; ++i0) {"
+  "  __global const float* ch0Row = (__global const float*)(imgCh0 + (ic0step * (i + i0) + ic0off + j * sizeof(float)));"
+  "  for (int j0 = 0; j0 < patchRadius; ++j0)"
+  "    patch[i0][j0] = ch0Row[j0];"
+  "}"
+  "const double pi = " STR_EXPAND(CV_PI) ";\n"
+  "#pragma unroll\n"
+  "for (int n0 = 0; n0 < 4; ++n0) {\n"
+  "#pragma unroll\n"
+  "  for (int n1 = 0; n1 < 4; ++n1) {"
+  "    double sum = 0;"
+  "    for (int i0 = 0; i0 < patchRadius; ++i0)"
+  "      for (int j0 = 0; j0 < patchRadius; ++j0)"
+  "        sum += patch[i0][j0] * cos(pi * (i0 + 0.5) * n0 / patchRadius) * cos(pi * (j0 + 0.5) * n1 / patchRadius);"
+  "    desc[n0 * 4 + n1] = sum / PR;"
+  "  }"
+  "}"
+  "for (int k = 0; k < 4; ++k) {"
+  "  desc[k] *= " STR_EXPAND(SQRT2_INV) ";"
+  "  desc[k * 4] *= " STR_EXPAND(SQRT2_INV) ";"
+  "}"
+  "double sum = 0;"
+  "for (int i0 = 0; i0 < patchRadius; ++i0) {"
+  "  __global const float* ch1Row = (__global const float*)(imgCh1 + (ic1step * (i + i0) + ic1off + j * sizeof(float)));"
+  "  for (int j0 = 0; j0 < patchRadius; ++j0)"
+  "    sum += ch1Row[j0];"
+  "}"
+  "desc[16] = sum / patchRadius;"
+  "sum = 0;"
+  "for (int i0 = 0; i0 < patchRadius; ++i0) {"
+  "  __global const float* ch2Row = (__global const float*)(imgCh2 + (ic2step * (i + i0) + ic2off + j * sizeof(float)));"
+  "  for (int j0 = 0; j0 < patchRadius; ++j0)"
+  "    sum += ch2Row[j0];"
+  "}"
+  "desc[17] = sum / patchRadius;"
+  "}" );
+
+#undef STR_EXPAND_
+#undef STR_EXPAND
+
+class ParallelDCTFiller : public ParallelLoopBody
+{
+private:
+  const Size sz;
+  const Mat *imgCh;
+  std::vector< GPCPatchDescriptor > *descr;
+
+  ParallelDCTFiller &operator=( const ParallelDCTFiller & );
+
+public:
+  ParallelDCTFiller( const Size &_sz, const Mat *_imgCh, std::vector< GPCPatchDescriptor > *_descr )
+      : sz( _sz ), imgCh( _imgCh ), descr( _descr ){};
+
+  void operator()( const Range &range ) const
+  {
+    for ( int i = range.start; i < range.end; ++i )
+    {
+      int x, y;
+      GPCDetails::getCoordinatesFromIndex( i, sz, x, y );
+      getDCTPatchDescriptor( descr->at( i ), imgCh, y, x );
+    }
+  }
+};
+
+void getAllDCTDescriptorsForImage( const Mat *imgCh, std::vector< GPCPatchDescriptor > &descr, const GPCMatchingParams &mp )
+{
+  const Size sz = imgCh[0].size();
+  descr.reserve( ( sz.height - 2 * patchRadius ) * ( sz.width - 2 * patchRadius ) );
+
+  if ( mp.useOpenCL && ocl::useOpenCL() )
```

Comment:
It is better to extract OpenCL specific code to separate function and use `CV_OCL_RUN` macro and `#ifdef HAVE_OPENCL` guards. Please take a look at the [example](https://github.com/opencv/opencv/blob/master/modules/imgproc/src/blend.cpp#L94-L151).

Also I think it is better to remove `GPCMatchingParams::useOpenCL` field and use only global `ocl::useOpenCL()` function (+ check whether the input is UMat).